File: lib/engine/translate.dart
// 自由文本翻译：MyMemory 免费 API + 缓存 + 静默降级（纯 Dart）
import 'dart:convert';

import 'package:http/http.dart' as http;

final Map<String, List<Object>> _cache = {}; // key -> [expiryMs, text]
const _ttlOk = 24 * 3600 * 1000;
const _ttlFail = 10 * 60 * 1000;

Future<String> translateZhTo(String text, String target) async {
  if (text.isEmpty) return '';
  final key = '$text|$target';
  final now = DateTime.now().millisecondsSinceEpoch;
  final hit = _cache[key];
  if (hit != null && (hit[0] as int) > now) return hit[1] as String;
  try {
    final r = await http
        .get(Uri.https('api.mymemory.translated.net', '/get',
            {'q': text, 'langpair': 'zh-CN|$target'}))
        .timeout(const Duration(seconds: 6));
    if (r.statusCode == 200) {
      final d = jsonDecode(utf8.decode(r.bodyBytes)) as Map<String, dynamic>;
      final out = ((d['responseData'] ?? {}) as Map)['translatedText'] ?? '';
      if (out is String && out.isNotEmpty && !out.toUpperCase().contains('MYMEMORY WARNING')) {
        _cache[key] = [now + _ttlOk, out];
        return out;
      }
    }
  } catch (_) {}
  _cache[key] = [now + _ttlFail, ''];
  return '';
}

File: backend/app/translate.py
"""自由文本翻译：MyMemory 免费 API（无 key），带缓存与静默降级。

生物分类词的多语言映射由 taxonomy 静态词典承担（精确、零网络依赖），
这里只处理用户的自由文本（如「雪山杀人狂」）到目标语言的转换。
"""
from __future__ import annotations

import time

import httpx

_cache: dict[tuple, tuple[float, str]] = {}
_TTL_OK = 24 * 3600
_TTL_FAIL = 600

LANG_TARGET = {"en": "en", "ja": "ja", "ko": "ko"}


async def translate(text: str, target: str, source: str = "zh-CN") -> str:
    if not text or target not in LANG_TARGET:
        return ""
    key = (text, source, target)
    now = time.time()
    if key in _cache:
        until, val = _cache[key]
        if until > now:
            return val
    try:
        async with httpx.AsyncClient(timeout=6.0) as client:
            r = await client.get(
                "https://api.mymemory.translated.net/get",
                params={"q": text, "langpair": f"{source}|{LANG_TARGET[target]}"},
            )
            out = ((r.json().get("responseData") or {}).get("translatedText")) or ""
            if out and "MYMEMORY WARNING" not in out.upper():
                _cache[key] = (now + _TTL_OK, out)
                return out
    except Exception:
        pass
    _cache[key] = (now + _TTL_FAIL, "")
    return ""
